src/utils.jl

```diff
@@ -17,6 +17,6 @@ ofeltype(x, y) = convert(float(eltype(x)), y)
 
 _eps(T::Type{<:AbstractFloat}, e) = T(e)
 # catch complex and integers
```

Comment:
Maybe we don't need the first method at all. But could add a docstring?
```suggestion
"""
    _eps(Type{T}, val)

Mostly this produces `real(T)(val)`, so that `_eps(Float32, 1e-8) === 1f-8` will
convert the Float64 parameter epsilon to work nicely with Float32 parameter arrays.

But for Float16, it imposes a minimum of `Float16(1e-7)`, unless `val==0`.
This is basically a hack to increase the default epsilon, to help many optimisers avoid NaN.
"""
```